File: YTTabBarControllerDemo(OC)/YTTabBarControllerDemo/YTTabBarControler/YTTabBarController.m
//
//  YTTabBarController.m
//  02-自定制TabBar
//
//  Created by 余婷 on 16/6/4.
//  Copyright © 2016年 余婷. All rights reserved.
//

#import "YTTabBarController.h"
#import "YTTabBar.h"
#import "YTTabBarButton.h"

//遵守协议
@interface YTTabBarController ()<YTTabBarDelegate>

@property(nonatomic, strong)YTTabBar * yt_tabBar;

@end

@implementation YTTabBarController


- (void)viewDidAppear:(BOOL)animated{

    [super viewDidAppear:animated];
    
    //1.删除自动创建的tabBarbutton
    for (UIView * view in self.tabBar.subviews) {
        [view removeFromSuperview];
    }
    
    
    
    [self.tabBar addSubview:self.yt_tabBar];
    
    //2.设置选中下标
    self.yt_tabBar.selectedIndex = self.selectedIndex;
    
    self.yt_tabBar.backgroundImage = [UIImage imageNamed:@"tabBar"];
   
    
    
}
#pragma mark - 懒加载
- (YTTabBar *)yt_tabBar{

    if (_yt_tabBar == nil) {
        
        _yt_tabBar = [[YTTabBar alloc] init];
        _yt_tabBar.frame = self.tabBar.bounds;
        _yt_tabBar.backgroundColor = [UIColor whiteColor];
        //设置代理
        _yt_tabBar.delegate = self;
        
        
    }
    return _yt_tabBar;
}

#pragma mark - 添加子视图控制器
- (void)addController:(UIViewController *)controller WithTitle:(NSString *)title imageName:(NSString *)imageName selectImageName:(NSString *)selectImageName{

    //设置tabBarItem
    controller.tabBarItem.title = title;
    controller.tabBarItem.image = [UIImage imageNamed:imageName];
    controller.tabBarItem.selectedImage = [[UIImage imageNamed:selectImageName] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
    
    //将视图控制器添加标签栏控制器中
    [self addChildViewController:controller];
    
    
    
    //可以让自己定制的tabBar去创建一个对应的按钮
    [self.yt_tabBar addbuttonWithItem:controller.tabBarItem];
    
}

#pragma mark - YTTabBar Delegate
- (void)changeControllerWithIndex:(NSInteger)index{

    //切换视图控制器
    self.selectedIndex = index;
}

#pragma mark - 外部给centerView赋值
- (void)setCenterView:(UIView *)centerView{

    _centerView = centerView;
    
    self.yt_tabBar.centerView = centerView;
}





@end

File: YTTabBarController(Swift)/YTTabBarController.swift
//
//  YTTabBarController.swift
//  01-自定制tabBarController
//
//  Created by 余婷 on 16/9/10.
//  Copyright © 2016年 余婷. All rights reserved.
//

import UIKit

class YTTabBarController: UITabBarController {
    
    //MARK: - 属性
    lazy var yt_tabBar:YTTabBar = {
    
        //自己定义的tabBar的大小和系统自带的tabBar的大小一样
        let tempTabBar = YTTabBar(frame:self.tabBar.bounds)
        
        //点击tabBar上的按钮去切换视图控制器
        tempTabBar.changeSelectedIndex = {(index)in
        
            //切换到指定的视图控制器
            self.selectedIndex = index
        }
        
        //将自定义的tabBar贴到系统的tabBar上
        self.tabBar.addSubview(tempTabBar)
        
        
        return tempTabBar
    }()

    //MARK: - 生命周期
    override func viewDidLoad() {
        super.viewDidLoad()
        
       

    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        //设置默认选中的按钮
        self.yt_tabBar.selectedIndex = self.selectedIndex
        
        //改变文字颜色
        self.yt_tabBar.titleColor = UIColor.lightGrayColor()
        
        //移除tabBar上自动添加的所有的子视图
        for item in self.tabBar.subviews {
            //判断子视图的类型是否是YTTabBar
            //isKindOfClass判断指定的对象是否是指定的类型
            if item.isKindOfClass(YTTabBar.self) {
                continue
            }
            
            //将不是YTTabBar的从系统的tabBar上移除
            item.removeFromSuperview()
            
        }
        
    }
    
    
}

//MARK: - 添加子视图控制器
extension YTTabBarController{
    
    ///添加子视图控制器
    func addController(controller:UIViewController,title:String,imageName:String,selectImageName:String){
        
        //设置视图控制器对一个的tabBarItem
        controller.tabBarItem.title = title
        controller.tabBarItem.image = UIImage.init(named: imageName)?.imageWithRenderingMode(.AlwaysOriginal)
        controller.tabBarItem.selectedImage = UIImage.init(named: selectImageName)?.imageWithRenderingMode(.AlwaysOriginal)
        //将视图控制器添加到tabBarController中
        self.addChildViewController(controller)
        
        //在yt_tabBar上创建对应按钮
        self.yt_tabBar.addButtonWithItem(controller.tabBarItem)
        
    }
}
